src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -370,61 +450,81 @@ public TStructuralType CopyChangedValues(TStructuralType original)
+                throw Error.ArgumentNull(nameof(original));
+            }
 
-                    _allProperties[nestedResourceName].SetValue(original, instance);
-                }
-                else
-                {
-                    // Recursively patch the subtree.                    
-                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+            if (!_structuredType.IsInstanceOfType(original))
+            {
+                throw Error.Argument(nameof(original), SRResources.DeltaTypeMismatch, _structuredType, original.GetType());
+            }
 
-                    Type newType = deltaNestedResource.StructuredType;
-                    Type originalType = originalNestedResource.GetType();
+            IEnumerable<PropertyAccessor<TStructuralType>> propertiesToCopy = GetUnchangedPropertyNames().Select(s => _allProperties[s]);
+            foreach (PropertyAccessor<TStructuralType> propertyToCopy in propertiesToCopy)
+            {
+                propertyToCopy.Copy(_instance, original);
+            }
 
-                    if (deltaNestedResource.IsComplexType && newType != originalType)
-                    {
-                        originalNestedResource = ReAssignComplexDerivedType(originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
-                        _structuredType.GetProperty(nestedResourceName).SetValue(original, (object)originalNestedResource);
-                    }
+            CopyUnchangedDynamicValues(original);
+        }
 
-                    deltaNestedResource.CopyChangedValues(originalNestedResource);
-                }
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the changes tracked by this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PATCH operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public TStructuralType Patch(TStructuralType original)
+        {
+            if (IsComplexType)
+            {
+                original = ReassignComplexDerivedType(original, _structuredType, original.GetType(), ExpectedClrType) as TStructuralType;
             }
 
-            return original;
+            return CopyChangedValues(original);
         }
 
-        private static dynamic ReAssignComplexDerivedType(dynamic originalValue, Type newType, Type originalType, Type declaredType)
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the values stored in this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PUT operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public void Put(TStructuralType original)
```

Comment:
I think this is the reason S had changed the return type of `Patch` to `void` but it will be a breaking change for `patch` and maybe this is something that can be done in a major release.

Reply:
@habbes Think about it this way.
I have a `customer` object with a property `BillingAddress` declared as `Address` the base type of `PostalAddress` and `StreetAddress`. Suppose the `BillingAddress` property is storing a `PostalAddress` and then we try to patch the property with a `StreetAddress`. The `Patch` method would need to return the patched object - which is basically a reference to a new `StreetAddress` instance.
Below is how the code would look like:
```csharp
var customer = customers.SingleOrDefault(d => d.Id == key); // At this point, customer.BillingAddress is a PostalAddress
var patched = delta.Patch(customer.BillingAddress); // This call with leave BillingAddress unchanged but the returned value is a StreetAddress
customer.BillingAddress = patched; // We need to do this to swap the PostalAddress with the StreetAddress
```

If however we had a variant of `Patch` method that accepted the entity that the complex type property belonged to, we could make the method return `void` since we'd able to do the last step - swapping the property value - from within the `Patch` method. In the current implementation, the type of parameter passed to `Patch` method (defined in `DeltaOfTStructuredType` class) is expected to be assignable to `TStructuredType` in `Patch(Delta<TStructuredType> delta)` controller action. I'm assuming this limitation/challenge is why Sr changed the `Patch` method to return the patched object